### ISSUE TITLE
ParamSpec cannot capture generic function param in __call__

### ISSUE THREAD

**Issue by lambyqq** [microsoft/pyright#5033]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
I'm writing a function decorator as callable class that uses ParamSpec to capture the function's param types.  However, `pyright` does not correctly resolve param type when `func` itself is generic:

```
class GenericCallableWrapper(Generic[P, T]):
    def __init__(self, func: Callable[P, T]) -> None:
        self._func = func

    def __call__(self, *__args: P.args, **__kwargs: P.kwargs) -> T:
        return self._func(*__args, **__kwargs)

@GenericCallableWrapper
def identify_func(t: T) -> T:
    return t

ret = identify_func('str value')
reveal_type(ret)
```
This would raise an error:
```
error: Argument of type "Literal['str value']" cannot be assigned to parameter "t" of type "T@identify_func" in function "__call__"
```
and revealed type is still generic: `information: Type of "ret" is "T@identify_func"`

**Expected behavior**
pyright should resolve the type of `ret` correctly as `str` in this case.

**VS Code extension or command-line**
pyright version: 1.1.305
Run both pylance (vs code extension) as well as command line pyright
Python 3.8.13

**Additional context**
Interestingly, when I add a `def func() -> Callable[P, T]` in `GenericCallableWrapper`, then
```
    ret = identify_func.func()('str value')
    reveal_type(ret)
```
will correctly resolve the type of `ret` as str

**Comment by erictraut** [comment]
The problem here is that `T@identify_func` is no longer in scope once it is captured by the `GenericCallableWrapper` class. TypeVars must be solved within the context of the scope in which they are defined. When a TypeVar is no longer in scope, it cannot be solved. In this case, the scope for `T@identify_func` is the undecorated `identify_func`, so it must be solved by calling this function directly. Once you wrap it and capture the TypeVar in the `GenericCallableWrapper`, the TypeVar is no longer in scope.

Pyright does contain special-case handling for a decorator function that returns a `Callable` type that allows out-of-scope TypeVars to be solved as though they belonged to the `Callable`'s own scope. For example, this works:

```python
def GenericCallableWrapper(func: Callable[P, T]) -> Callable[P, T]:
    return func
```

But this special-case handling doesn't extend to generic classes like `GenericCallableWrapper` in your sample code above.

It appears that mypy provides special-case handling for methods that include an `*args` and `**kwargs` parameter with an associated ParamSpec. I'll need to think more about whether it makes sense to add such a special case in pyright as well.

**Comment by erictraut** [comment]
I've implemented support for this in a manner that is consistent with mypy. I think this is a case where practicality beats purity, and it should be possible for these no-longer-in-scope TypeVars to be solved in the context of the `__call__` method.

This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.306, which I just published. It will also be included in a future release of pylance.

**Comment by stavalfi** [comment]
any update? is it a pylanc problem?

**Comment by erictraut** [comment]
@stavalfi, this issue was fixed and closed over a year ago. If you have questions, feel free to post to the discussion section. If you think that you've identified a bug, please file a new issue.
